src/dhcpmon/src/dhcp_device.c

```diff
@@ -0,0 +1,435 @@
+                break;
+            default:
+                break;
+            }
+
+            if (stop_dhcp_rpocessing == 1) {
+                break;
+            }
+
+            if (dhcp_option[offset] == 0) { // DHCP Option Padding
+                offset++;
+            } else {
+                offset += dhcp_option[offset + 1] + 2;
+            }
+        }
+    } else {
+        syslog(LOG_ALERT, "pcap_callback(%s): snap length is too low to capture DHCP options", context->intf);
+        exit(EXIT_FAILURE);
+    }
+}
+
+/**
+ * @code dhcp_device_process(arg);
+ *
+ * @brief validate current interface counters by comparing aggregate counter with snapshot counters.
+ *
+ * @param arg   user provided argument for capturing process (interface context)
+ *
+ * @return none
+ */
+static void* dhcp_device_process(void *arg)
+{
+    dhcp_device_context_t *context = (dhcp_device_context_t*) arg;
+
+    if (pcap_loop(context->cap, -1, pcap_callback, arg) < 0) {
+        syslog(LOG_ALERT, "pcap_loop(%s): %s", context->intf, pcap_geterr(context->cap));
+        exit(EXIT_FAILURE);
+    }
+
+    pthread_exit(NULL);
+}
+
+/**
+ * @code dhcp_device_validate(counters, counters_snapshot);
+ *
+ * @brief validate current interface counters by comparing aggregate counter with snapshot counters.
+ *
+ * @param counters              recent interface counter
+ * @param counters_snapshot     snapshot counters
+ *
+ * @return DHCP_MON_STATUS_HEALTHY, DHCP_MON_STATUS_UNHEALTHY, or DHCP_MON_STATUS_INDETERMINATE
+ */
+static dhcp_mon_status_t dhcp_device_validate(dhcp_device_counters_t *counters,
+                                              dhcp_device_counters_t *counters_snapshot)
+{
+    dhcp_mon_status_t rv = DHCP_MON_STATUS_HEALTHY;
+
+    if ((counters[DHCP_RX].discovery == counters_snapshot[DHCP_RX].discovery) &&
+        (counters[DHCP_RX].offer == counters_snapshot[DHCP_RX].offer) &&
+        (counters[DHCP_RX].request == counters_snapshot[DHCP_RX].request) &&
+        (counters[DHCP_RX].ack == counters_snapshot[DHCP_RX].ack)) {
+        rv = DHCP_MON_STATUS_INDETERMINATE;
+    } else {
+        // if we have rx DORA then we should have corresponding tx DORA (DORA being relayed)
+        if (((counters[DHCP_RX].discovery > counters_snapshot[DHCP_RX].discovery ) &&
+             (counters[DHCP_TX].discovery <= counters_snapshot[DHCP_TX].discovery)    ) ||
+            ((counters[DHCP_RX].offer > counters_snapshot[DHCP_RX].offer ) &&
+             (counters[DHCP_TX].offer <= counters_snapshot[DHCP_TX].offer)            ) ||
+            ((counters[DHCP_RX].request > counters_snapshot[DHCP_RX].request ) &&
+             (counters[DHCP_TX].request <= counters_snapshot[DHCP_TX].request)        ) ||
+            ((counters[DHCP_RX].ack > counters_snapshot[DHCP_RX].ack ) &&
+             (counters[DHCP_TX].ack <= counters_snapshot[DHCP_TX].ack)                )) {
+            rv = DHCP_MON_STATUS_UNHEALTHY;
+        }
+    }
+
+    return rv;
+}
```

Comment:
What if we overflow the counters? Will that work?

Reply:
Counter won't overflow as these are 64-bit counters. Assuming a packet very 1usec, it would take about ~580k years to wrap the counter around (2^64÷(10^6×60×60×24×365))